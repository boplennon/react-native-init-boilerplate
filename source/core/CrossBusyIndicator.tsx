import React from 'react';
import { View, Text, TouchableOpacity } from 'react-native';
import {
  BallIndicator,
  BarIndicator,
  DotIndicator,
  MaterialIndicator,
  PacmanIndicator,
  PulseIndicator,
  SkypeIndicator,
  UIActivityIndicator,
  WaveIndicator,
} from 'react-native-indicators';

import styles, { Colors } from '../styles';

/**
 * Type of indicator
 *
 * @export
 * @enum {string}
 */
export enum IndicatorType {
  BallIndicator = 'BallIndicator',
  BarIndicator = 'BarIndicator',
  DotIndicator = 'DotIndicator',
  MaterialIndicator = 'MaterialIndicator',
  PacmanIndicator = 'PacmanIndicator',
  PulseIndicator = 'PulseIndicator',
  SkypeIndicator = 'SkypeIndicator',
  UIActivityIndicator = 'UIActivityIndicator',
  WaveIndicator = 'WaveIndicator',
}

/**
 * Generates a spinner of the provided type
 * @param param0 type and style
 */
export const Spinner = ({
  type,
  style = null,
  color = Colors.CrossLightBlue,
}: {
  type: IndicatorType;
  style?: any;
  color?: string;
}) => {
  switch (type) {
    case 'BallIndicator':
      return (
        <BallIndicator
          // @ts-ignore
          style={style}
          color={color}
        />
      );
    case 'BarIndicator':
      return (
        <BarIndicator
          // @ts-ignore
          style={style}
          color={color}
        />
      );
    case 'DotIndicator':
      return (
        <DotIndicator
          // @ts-ignore
          style={style}
          color={color}
        />
      );
    case 'PacmanIndicator':
      return (
        <PacmanIndicator
          // @ts-ignore
          style={style}
          color={color}
        />
      );
    case 'PulseIndicator':
      return (
        <PulseIndicator
          // @ts-ignore
          style={style}
          color={color}
        />
      );
    case 'SkypeIndicator':
      return (
        <SkypeIndicator
          // @ts-ignore
          style={style}
          color={color}
        />
      );
    case 'UIActivityIndicator':
      return (
        <UIActivityIndicator
          // @ts-ignore
          style={style}
          color={color}
        />
      );
    case 'WaveIndicator':
      return (
        <WaveIndicator
          // @ts-ignore
          style={style}
          color={color}
        />
      );
    default:
      return (
        <MaterialIndicator
          // @ts-ignore
          style={style}
          color={color}
        />
      );
  }
};

/**
 * Describes the props BusyIndicator accepts
 *
 * @export
 * @interface IBusyIndicatorProps
 */
export interface IBusyIndicatorProps {
  /**
   * Determines if the indicator is animating or not. Default value is true
   *
   * @type {boolean}
   * @memberof IBusyIndicatorProps
   */
  isBusy: boolean;
  /**
   * Optional message to show while busy. Defaults to empty string.
   *
   * @type {string}
   * @memberof IBusyIndicatorProps
   */
  message?: string;
  /**
   * Type of spinner to use
   *
   * @type {IndicatorType}
   * @memberof IBusyIndicatorProps
   */
  type?: IndicatorType;
  /**
   * Test id to assign this component
   *
   * @type {string}
   * @memberof IBusyIndicatorProps
   */
  testID?: string;
  /**
   * User pressed cancel
   */
  onCancel?: () => void;
}

/**
 * Displays a busy indicator (spinner) when the parameter is true
 * @param {bool} isBusy determines if the indicator is animating or not. Default value is true
 * @param {string} message optional message to show while busy. Defaults to empty string.
 * @param {IndicatorType} type of spinner.Defaults to "MaterialIndicator"
 */
const CrossBusyIndicator = ({
  isBusy = true,
  message = '',
  type = IndicatorType.MaterialIndicator,
  testID = '1',
  onCancel,
}: IBusyIndicatorProps) =>
  isBusy ? (
    <View style={styles.overlay} testID={testID}>
      <View style={[styles.absoluteCentered]}>
        <View style={[styles.columnContentTopCenter]}>
          <Spinner type={type} style={[styles.spinner]} />
        </View>
        <View style={styles.columnContentTopCenter}>
          <Text style={styles.textSpinner}>{message}</Text>
          {onCancel ? (
            <TouchableOpacity onPress={onCancel}>
              <Text style={[styles.textSpinner, { color: Colors.CancelButton }]}>
                Avbryt
              </Text>
            </TouchableOpacity>
          ) : null}
        </View>
      </View>
    </View>
  ) : null;

export default CrossBusyIndicator;
