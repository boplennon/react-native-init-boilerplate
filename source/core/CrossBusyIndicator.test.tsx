// core/components/BusyIndicator.test
/// <reference types="jest"/>
import React from 'react';
import TestRenderer from 'react-test-renderer';

import CrossBusyIndicator, {
  IBusyIndicatorProps,
  IndicatorType,
  Spinner,
} from './CrossBusyIndicator';

jest.unmock('react-native');

function setup(type = IndicatorType.MaterialIndicator) {
  const props: IBusyIndicatorProps = {
    isBusy: true,
    type,
  };

  const wrapper = TestRenderer.create(<CrossBusyIndicator {...props} />);

  return {
    props,
    wrapper,
  };
}

describe('components', () => {
  /**
   * Test component rendering. Properties of children might be tested by importing their type:
   *
   *    const busyProps = enzymeWrapper.find(BusyIndicator).props();
   *    expect(busyProps.isBusy).toBe(false);
   */
  describe('<CrossBusyIndicator />', () => {
    it('should render self and subcomponents', () => {
      const { wrapper } = setup();
      expect(wrapper.toJSON()).toMatchSnapshot();
    });

    it('should use DotIndicator type', () => {
      const { wrapper } = setup(IndicatorType.DotIndicator);
      expect(wrapper.root.findByType(Spinner)).toBeDefined();
    });
  });
});
