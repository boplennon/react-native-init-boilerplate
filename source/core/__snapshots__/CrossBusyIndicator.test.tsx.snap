// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components CrossBusyIndicator should render self and subcomponents 1`] = `
<View
  style={
    Object {
      "backgroundColor": "rgba(254, 222, 0, 0.3)",
      "bottom": 0,
      "left": 0,
      "position": "absolute",
      "right": 0,
      "top": 0,
    }
  }
  testID="1"
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "bottom": 0,
          "justifyContent": "center",
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        },
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "flex": 1,
            "flexDirection": "column",
            "justifyContent": "flex-start",
          },
        ]
      }
    >
      <View
        color="#057dc2"
        style={
          Array [
            Object {
              "alignItems": "flex-end",
              "flex": 1,
              "flexDirection": "row",
              "height": 50,
              "justifyContent": "center",
              "opacity": 1,
              "width": 50,
            },
          ]
        }
      />
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "flex": 1,
          "flexDirection": "column",
          "justifyContent": "flex-start",
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#057dc2",
            "fontSize": 20,
            "fontWeight": "normal",
            "marginTop": 5,
            "opacity": 1,
            "textAlign": "center",
          }
        }
      >
        
      </Text>
    </View>
  </View>
</View>
`;
